### ISSUE TITLE
xUnit1010 does not report invalid value for nullable uint in theory

### ISSUE THREAD

**Issue by dlemstra** [xunit/xunit#2983]
The xUnit1010 analyzer is currently not reporting that the value is invalid for the following theory:

```C# 
[Theory]
[InlineData(1)]
public void Test(uint? value)
{
    Assert.Equal(1U, value);
}
```

It looks like a check is missing. Also happens with the following code:

```C#
[Theory]
[InlineData(1U)]
public void Test(int? value)
{
    Assert.Equal(1, value);
}
```

**Comment by NormanFrieman** [comment]
Question:

Won't the CompatibleNumericValue_NullableType test give an error in these cases?
42 => uint?
42u => int?

Another question is, would it be interesting to have a warning if the value passed in the inlinedata is negative and the parameter is uint?

**Comment by KasinathCR** [comment]
@dlemstra - Shall I look into this?

**Comment by dlemstra** [comment]
You don't need to ask my permission to look into this. Feel free to take a look at how this can been solved.

**Comment by NormanFrieman** [comment]
@bradwilson 
While I was trying to find a way to solve this issue I had the following problem

I made a solution to the issue, but many tests that validate the conversion to null types broke, I'll post some examples here

![image](https://github.com/user-attachments/assets/c01f5d33-ff17-44c8-b699-a660566a5f40)
Before the fix, this test passed, but after the fix, this test (same scenario shown at the beginning of the issue) fails, because it expects the analyzer to find no error

![image](https://github.com/user-attachments/assets/a63f0fb1-8024-49dc-8076-5a47651dbf09)
Several other scenarios that used to pass are starting to give errors too

![image](https://github.com/user-attachments/assets/4e772561-f142-404d-ac07-09394345ba58)
In these scenarios, a runtime error also occurs

In all, more than 200 tests fail

This gave me a question: was this scenario really supposed to fail?

**Comment by bradwilson** [comment]
@NormanFrieman In those examples, I don't see any reason for failures. A value of `1` is compatible with `uint`, `ulong`, `ushort`, and `byte`. The core issue is that an _incompatible_ value is caught at runtime but not by the analyzer (for example, trying to pass `-1` to any unsigned value, or trying to pass a number which is too large like `ushort.MaxValue` being passed to `short`).

**Comment by bradwilson** [comment]
Unless I'm misunderstanding what you're saying :)

**Comment by bradwilson** [comment]
Oh, I see, the problem here isn't `int` => `uint`, it's `int` => `uint?` which suggests that the problem is with conversion when nullable is involved. I was confusing this issue with the open PR related to signed vs. unsigned.

![image](https://github.com/user-attachments/assets/b15778d5-9da1-4d1c-83fb-b23b2764ddfc)

**Comment by bradwilson** [comment]
FWIW, we don't consider this a runtime issue because we are relying on the framework to do the numeric conversions. Even though the compiler would happily coerce things, the runtime won't, and we're just reflecting the runtime's behavior here. In general we try our best to avoid re-implementing the compiler. 😂 

![image](https://github.com/user-attachments/assets/26e89520-46cc-482e-9678-5d5bb81ca6ce)

**Comment by NormanFrieman** [comment]
@bradwilson So what should be done in this case?

**Comment by bradwilson** [comment]
For this specific issue, the logic would look like:

- If the value is an integral numeric
- If the matching parameter is a nullable integral numeric
- Then raise xUnit1010 if the types aren't an exact match

From a test perspective:

* `int` => `int` is fine
* `int` => `int?` is fine
* `int` => `uint` is fine (assuming it fits)
* `int` => `uint?` is not fine

Integral numeric definition: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/integral-numeric-types

**Comment by NormanFrieman** [comment]
right, I'll add this logic and fix the tests in this [pr 189](https://github.com/xunit/xunit.analyzers/pull/189)

**Comment by bradwilson** [comment]
Closing as this was resolved in the above-linked PR.
